### PR TITLE
001 view job roles

### PR DESCRIPTION
**Pull request by RaggedyRach**
the backend api for the view job roles ticket

### CODE REVIEW

**Review by hannahmorgan24 — APPROVED** [review]

File: src/main/java/org/example/mappers/JobRoleMapper.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import java.util.List;
+
+public class JobRoleMapper {
+   private JobRoleMapper jobRoleMapper;
+    /*
+    maps the JobRoleResponse to a stream
+    and then uses Collectors to convert the stream to a List
+     */
+    public static List<OpenJobRoleResponse> mapJobRoleResponseList(
+            final List<OpenJobRoleResponse> jobRoleResponseList) {
```
Comment (shaunganley) [inline src/main/java/org/example/mappers/JobRoleMapper.java]: I don't think this does anything as you're mapping to the same type, you can just remove this class

File: src/main/java/org/example/models/Band.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+     */
+    BAND2("30,001-36,000"),
+    /**
+     * Enum of Band3 £36,001-£42,000 available for jobRoles.
+     */
+    BAND3("36,001-42,000"),
+    /**
+     * Enum of Band4 £42,001-£60,000 available for jobRoles.
+     */
+    BAND4("42,001-60,000");
```
Comment (shaunganley) [inline src/main/java/org/example/models/Band.java]: I don't think bands should link to salary, a band should be trainee, associate, senior associate, etc. Have a look at the useful links on the sprint board

File: src/main/java/org/example/models/Capability.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package org.example.models;
+
+public enum Capability {
```
Comment (shaunganley) [inline src/main/java/org/example/models/Capability.java]: making this an enum means we'll have to make a code change in the future if someone adds a new capability which we don't want to do. just use a string for capability

File: src/main/java/org/example/models/Band.java (hunk 2 above)
Comment (shaunganley) [reply src/main/java/org/example/models/Band.java]: also making this an enum means we'll have to make a code change in the future if someone adds a new capability which we don't want to do. just use a string for band

File: src/main/resources/db/migration/V1__create_jobRole_table.sql [hunk 4]
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE jobRole (
+    jobRoleID smallint PRIMARY KEY AUTO_INCREMENT NOT NULL,
+    jobRoleName varchar(100) NOT NULL,
+    jobRoleLocation ENUM ("BELFAST", "LONDON", "TORONTO", "BIRMINGHAM", "GDANSK"),
+	jobRoleCapability ENUM ("HR", "DELIVERY", "MARKETING", "SALES"),
+	jobRoleBand ENUM("BAND1","BAND2","BAND3","BAND4"),
```
Comment (shaunganley) [inline src/main/resources/db/migration/V1__create_jobRole_table.sql]: don't use enum for band or capability, create a separate table and link via foreign key. will allow us to create, read, update and delete them in future tickets

File: src/main/resources/db/migration/V4__create_prod_db.sql
Comment (shaunganley) [inline src/main/resources/db/migration/V4__create_prod_db.sql:1]: should this have been renamed?

File: src/test/java/org/example/controller/JobRoleControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    void getJobRoles_shouldReturnJobRolesResponseList_whenGetAllJobRolesReturnsJobRoleResponseList()
+            throws
+            SQLException, DatabaseConnectionException {
+        List<OpenJobRoleResponse> jobRoleResponseList = new ArrayList<>();
+
+        when(jobRoleService.getOpenJobRoles()).thenReturn(jobRoleResponseList);
+
+        Response response = jobRoleController.getJobRoles();
+
+        assertEquals(jobRoleResponseList, response.getEntity());
```
Comment (shaunganley) [inline src/test/java/org/example/controller/JobRoleControllerTest.java:45]: check 200 status code as well

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/main/resources/db/migration/V4__create_prod_db.sql
Comment (RaggedyRach) [reply src/main/resources/db/migration/V4__create_prod_db.sql:1]: ah! renamed the wrong one :( changed now.

File: src/main/java/org/example/mappers/JobRoleMapper.java (hunk 1 above)
Comment (RaggedyRach) [reply src/main/java/org/example/mappers/JobRoleMapper.java]: removed

File: src/test/java/org/example/controller/JobRoleControllerTest.java (hunk 5 above)
Comment (RaggedyRach) [reply src/test/java/org/example/controller/JobRoleControllerTest.java:45]: added additional assert statement to check 200 status

File: src/main/resources/db/migration/V5__create_jobRole_table.sql [hunk 6]
```diff
@@ -0,0 +1,19 @@
+CREATE TABLE jobRole (
+    jobRoleID smallint PRIMARY KEY AUTO_INCREMENT NOT NULL,
+    jobRoleName varchar(100) NOT NULL,
+    jobRoleLocation ENUM ("BELFAST", "LONDON", "TORONTO", "BIRMINGHAM", "GDANSK"),
+	capabiltyID smallint,
+	jobRoleBand varchar(100) NOT NULL,
```
Comment (shaunganley) [inline src/main/resources/db/migration/V5__create_jobRole_table.sql]: create a band table and link as you've done with capability

**Review by shaunganley — CHANGES_REQUESTED** [review]

**Review by shaunganley — APPROVED** [review]

**Review by gerardmclean2 — APPROVED** [review]

**Review by JakubRS7 — APPROVED** [review]

**Pull request merged**